0023-keymanager-secret-sharing.md

```diff
@@ -0,0 +1,671 @@
+  fn proactive_randomization(&self, churp_id: u8, round: u64);
+
+  /// Construct a full share from the switch data points obtained from 
+  /// the nodes in the new committee.
+  ///
+  /// Full share:
+  ///      s'(y) = B'(node_id,y)
+  fn full_share_distribution(&self, churp_id: u8, round: u64);
+
+  /// Return switch data point (integer) for the calling node.
+  /// 
+  /// The point is evaluation of the shared secret (polynomial)
+  /// at the given y value (caller id).
+  /// 
+  /// Switch point:
+  ///     P = g(caller_id) = B(node_id,caller_id)
+  ///
+  /// WARNING: This method must be called over a secure channel as
+  /// the polynomial needs to be kept secret and generated only
+  /// for authorized nodes.
+  fn switch_point(&self, churp_id: u8, round: u64) -> Integer;
+
+  /// Return the key share for the given round.
+  ///
+  /// To construct a derived key, the caller is required to collect 
+  /// at least the threshold number of key shares and combine them
+  /// locally to construct the derived key.
+  ///
+  /// Key share:
+  ///     K = H(runtime_id, key_pair_id)^{r g(0)}
+  ///
+  /// WARNING: This method must be called over a secure channel as
+  /// the key share needs to be kept secret and generated only
+  /// for authorized nodes.
+  fn key_share(&self, churp_id: u8, round: u64, runtime_id: Namespace, 
+    key_pair_id: KeyPairId) -> Integer;
+}
+```
+
+Methods:
+
+- `churp/init`
+
+- `churp/verification_matrix`
+
+- `churp/bivariate_share`
+
+- `churp/dealing`
+
+- `churp/share_reduction`
+
+- `churp/proactive_randomization`
+
+- `churp/full_share_distribution`
+
+- `churp/switch_point`
+
+- `churp/key_share`
+
+### Key manager client
+
+```rust
+/// Key manager client for CHURP.
+pub trait Client {
+  /// Return the key for the given round.
+  ///
+  /// The key is combined from key shares obtained from the key manager nodes
+  /// in the current CHURP committee.
+  fn key(&self, churp_id: u8, round: u64, runtime_id: Namespace, key_pair_id: KeyPairId) -> Vec<u8>;
+}
+```
+
+### Positive (CHURP)
+
+- High security, as the master secret is shared among key manager nodes.
+- Supports proactive randomization (share refresh).
+- Dynamic committees.
+
+### Negative (CHURP)
+
+- Slower than KDC.
+
+## Key Derivation Center
+
+Key Derivation Center (KDC) is a secret sharing scheme based on
+the verifiable secret sharing scheme (VSS) where every node possesses
+only a share of the master secret. To derive a key from the master secret,
+one needs to obtain at least t+1 threshold number of key shares from distinct
+nodes and reassemble them locally.
+
+```rust
+/// Key manager application that implements key derivation center (KDC).
+pub trait KDC {
+  /// TODO: Define methods if we decide to implement KDC also.
+}
+```
+
+### Positive (KDC)
+
+- High security, as the master secret is shared among key manager nodes.
+- Supports proactive randomization (share refresh).
+- Faster than CHURP.
```

Comment:
Any approximate numbers for comparison?

Reply:
This comment is not accurate, so I changed it.

The cost of issuing a derived-key share should be similar. The only difference is that in CHURP the dealer needs to evaluate one polynomial more (which is depended on the configuration, but should be negligible comparing to other operations). Handoffs are more complex, but can be done in parallel. So their complexity can be ignored.

According to Bennet's comment, key derivation should be cheap, about 150 μS. Not sure how long it takes to combine the shares.

Detection of corrupted shared on the client side is a different story. The client cannot verify each share separately against verification matrix/vector, however, he can use multiple shares to find a corrupted one.

Reply:
> Handoffs are more complex, but can be done in parallel. So their complexity can be ignored.
> Not sure how long it takes to combine the shares.

Latency numbers on both would be interesting to see since the prize is for contract state keys to be dispersed. For this to work, because reconstructing shares is a potentially lengthy operation, we will need to enable [access lists](https://eips.ethereum.org/EIPS/eip-2930) so that the keys for all contracts can be pre-fetched and avoid users needing to pay gas proportional to the slowness of the algorithm. It would take a not-insignificant amount of effort to enable access lists throughout the stack, so if either is fast enough not to need it, we're saving a lot of engineering time as well.

Is the purpose of this extension to support contract state or off-chain key management? If it's the latter, performance doesn't matter, but neither does trusting the dealer.